File: src/string/find_common_char.py
def find_common_char(string: str):
  common_char = string[0]
  common_char_count = string.count(common_char)
  cache = { "common": dict(char = common_char, count = common_char_count) } 
  cache[common_char] = common_char_count
  for char in string:
    if char in cache: continue
    char_count = string.count(char)
    cache[char] = char_count
    if char_count > cache["common"]["count"]:
      cache["common"]["char"] = char
      cache["common"]["count"] = char_count
  return cache["common"]

File: src/string/find_common_char.ts
export interface CharCount {
  char: string;
  count: number;
}

interface CommonCharCache {
  [letter: string]: CharCount;
  max: CharCount;
}

export default function findCommonChar(str: string): CharCount {
  const firstCharCount = getCharCount(str[0], str);
  const cache: CommonCharCache = { [str[0]]: firstCharCount, max: firstCharCount };

  for (let index = 0; index < str.length; index++) {
    const char = str[index];
    if (char in cache) continue;
    const count = (cache[char] = getCharCount(char, str));
    if (count.count > cache.max.count) cache.max = count;
  }

  return cache.max;
}

function getCharCount(char: string, str: string): CharCount {
  let count = 0;

  for (let index = 0; index < str.length; index++) {
    const letter = str[index];
    if (char === letter) count++;
  }

  return { char, count: count };
}
